Add top level rails application url

aka47 opened the pull request:
Fixes #39566

### Other Information

The http/request.rb object is a complicated one - with around 6 different modules mixed in - Rack::Helpers, Rack::Env and others. I renamed  ActionDispatch::Http::URL to ActionDispatch::Http::RequestURL as it is mainly building an URL out of the request object, for all the different settings and edge cases. Relying heavily on Rack in the background. Some of the methods, especially around constructing the port - need to stay part of RequestUrl as they are need to build the URL out of the request object and can not move to the URI class, unfortunately.

Once the url is build, we parse it and can work with it as an URI object. 

Also - with the next release of Rack, Rack::Request::Env will be gone and that needs to be reflected in ActionDispatch::Request Class then.

aka47 committed:
Add config.application_url and ActionDispatch::Http::URI class

Adding Rails.application.url
This adds config.application_url. It is available as Rails.application.url
If config.application_url is set, it is used to configure config.default_url_options and
config.action_mailer.default_url_options as well.

Adding ActionDispatch::Http::URI class
It uses Rubys Core URI class and adds some rails specific methods to it.
All class methods from ActionDispatch::Http::URL are moved over to URI
start simplifing it.

p8 commented:
@aka47 Thanks for contributing to Rails!
There also is a previous PR for this: https://github.com/rails/rails/pull/39629
@vipulnsward Are you still working on that one?

aka47 commented:
Thanks @p8. It's a start and I am happy to learn and add possible and more PR to rails in the future. Some parts of the PR from @vipulnsward could be merged into this one, setting the default-url-options for one. I will do that. 

It has 2 failing tests - which have the same root cause - I will fix those some time soon.

zzak commented:
Vipul said he would look into updating his PR soon, let's give him a chance to pick up where he left off 🙏

aka47 commented:
> Vipul said he would look into updating his PR soon, let's give him a chance to pick up where he left off 🙏

I am more than happy to work with @vipulnsward on this PR. Vipul - lets not do the same work twice, lets better work together on it? How about that?

marivaldo commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,156 @@
+            parts = host.split(".")
+            parts[0..-(tld_length + 2)]
+          end
+
```
```suggestion
```

marivaldo commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -49,6 +49,7 @@ class Railtie < Rails::Railtie #:nodoc:
       # Ensure readers methods get compiled.
       options.asset_host        ||= app.config.asset_host
       options.relative_url_root ||= app.config.relative_url_root
+      options.default_url_options ||= app.default_url_options
```
```suggestion
      options.asset_host          ||= app.config.asset_host
      options.relative_url_root   ||= app.config.relative_url_root
      options.default_url_options ||= app.default_url_options
```
can we align?

vipulnsward commented:
@ak15 you can take over if you are interested. I have been caught up in work, so might not get to it soon.

ak15 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,155 @@
+      attr :core_uri
+
+      delegate :scheme, :host, :port, :path, :query, :fragment, :to_s, to: :core_uri
+      delegate :scheme=, :host=, :port=, :path=, :query=, :fragment= , to: :core_uri
```
```suggestion
      delegate :scheme=, :host=, :port=, :path=, :query=, :fragment=, to: :core_uri
```


very small 😉

p8 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,155 @@
+module ActionDispatch
+  module Http
+    class URI
+      IP_HOST_REGEXP = /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/
```
This regexp also matches `example.com.1.2.3.4`
```suggestion
      IP_HOST_REGEXP = /\A\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\z/
```

marivaldo commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -325,24 +277,43 @@ def server_port
+
+      def request_protocol
+        @protocol ||= ssl? ? "https://" : "http://"
+      end
+
```
```suggestion
      end
```

aka47 commented:
What are the next steps here? @p8 / @marivaldo - please help me out here. I am happy to do some more work, tune it, but I need someone who gives me a path to its being merged. That would be wonderful. 

Working on this task I got the impression this could need some more follow up refactoring - as this part is very complex and interdependent. If this should be part of this PR - I would love to have someone with whom I could share my ideas for it and come an agreement together. 
It starts with moving ActionDispatch::Request outside of the HTTP folder to the action_dispatch folder. FilterParamenter could be moved to the request-folder. The content_security_policy could be moved middleware or the action_dispatch folder, but not sure. In general it would be great to declutter the http folder and bring a bit more clarity into the whole thing. ( It seems like http is such an open name - it invites everyone to put things into it.)

So - should we just bring this to a quick end and merge it? Or what is the plan here forward?  .. Thanks a lot ..

p8 commented on actionpack/test/dispatch/uri_test.rb:
```diff
@@ -0,0 +1,48 @@
+require "abstract_unit"
+
+class URITest < ActiveSupport::TestCase
+  def setup
+  end
```
If there is no setup I think it can be removed.

p8 commented on railties/lib/rails/application/bootstrap.rb:
```diff
@@ -88,6 +89,16 @@ module Bootstrap
+
+      # Sets the Rails.application.url config
+      initializer :initialize_url, group: :all do
+        if config.application_url
+          Rails.application.url = ActionDispatch::Http::URI.build_from_string(config.application_url)
+          if Rails.application.default_url_options.blank?
+            Rails.application.default_url_options = { host: Rails.application.url.host, protocol: Rails.application.url.scheme }
+          end
```
Maybe using a temp variable could make this a little more readable.
I also think the port is missing.
```suggestion
        if config.application_url
          url = ActionDispatch::Http::URI.build_from_string(config.application_url)
          Rails.application.url = url
          if Rails.application.default_url_options.blank?
            Rails.application.default_url_options = {
              host: url.host,
              protocol: url.scheme,
              port: url.port
            }
          end
```

Or maybe define a method on ActionDispatch::Http::URI so we can do something like:

```suggestion
        if config.application_url
          url = ActionDispatch::Http::URI.build_from_string(config.application_url)
          Rails.application.url = url
          if Rails.application.default_url_options.blank?
            Rails.application.default_url_options = url.default_options
          end
```

p8 replied:
I think calling `host_and_port` is nice as well:
https://github.com/rails/rails/pull/39629/files#diff-55bc5dde25c29798e0e8fceb70e7391a67a262e0cc2550d179eefa0a3bbc4cffR60

aka47 replied:
done

p8 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,152 @@
+        @core_uri = uri
+      end
+
+      def self.build_from_faulty_string(url_string)
```
What is the reason you want to handle faulty strings?

aka47 replied:
changed this ..

aka47 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,152 @@
+        else
+          # handle other edge cases
+          url_string = "" if url_string == ":"
+          new(::URI.parse(url_string)) rescue nil
```
true

aka47 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,152 @@
+  module Http
+    class URI
+      mattr_accessor :tld_length, default: 1
+      attr :core_uri
```
Maybe .. core_uri does reference that it is a real URI object from ruby .. not just a string. But it does not seem to be clear.. but original_uri also does not communicate this.

halo replied:
Suggestions: `upstream_uri`, `native_uri`, `vanilla_uri`, `basic_uri`. For what it's worth, I think `uri` and `original_uri` are good suggestions, too. When I read `core_uri` I thought this belongs to some Rails "core" code.

aka47 commented:
maybe @rafaelfranca you can have a lot at this PR and my questions around how to finalize it. Or you know someone. This PR is waiting for good feedback to get over the finish line for some time and if nothing happens - it will not move forward. 
I still have energie and time to finish it - and go from here to do some more work on rails. I would love to do that - just need some feedback here. Thanks a million times.

halo commented on railties/test/application/configuration_test.rb:
```diff
@@ -281,6 +281,20 @@ def change
+      assert_instance_of ActionDispatch::Http::URL, Rails.application.url
+      assert_equal "https://rails.test", Rails.application.url.to_s
+      assert_equal({ host: "rails.test", protocol: "https", port: 443 }, Rails.application.default_url_options)
+      assert_equal({ host: "rails.test", protocol: "https", port: 443 }, Rails.application.config.action_mailer.default_url_options)
```
I'm wondering how this list relates to `ApplicationController.renderer.defaults`. 

Would/should that also be affected?

```suggestion
      assert_equal({ host: "rails.test", protocol: "https", port: 443 }, Rails.application.config.action_mailer.default_url_options)
      assert_equal({ http_host: "rails.test", https: true }, ApplicationController.renderer.defaults)
Write
```

supairish replied:
Every time i've used a controller renderer outside of the request/response cycle it's never appeared to inherit the default url options from anywhere they're defined in the app.  I've always had to manually assign 'http_host' and 'https'.  Would be awesome if this PR addressed that!

aka47 replied:
That does not work yet, I have never worked with ApplicationController.renderer.defaults. How did you use it, what was the problem and how did you solve it?

halo replied:
Hi! I **just** noticed, that this seems to have been addressed [just a few months ago](https://github.com/rails/rails/commit/93038baad70d477f6cbb06fa9d48c8e429fb192f).

The problem was, for me, that 

```ruby
ApplicationController.renderer.render inline: "<%= image_tag user.avatar %>"
```

would not inherit the host and protocol from `Rails.application.default_url_options` when rendered outside of a request cycle. 

A solution, so far, had been to add this in an initializer:

```ruby
ActiveSupport::Reloader.to_prepare do
  default_url_options = Rails.application.config.action_mailer.default_url_options

  ApplicationController.renderer.defaults.merge!(
    http_host: default_url_options[:host],
    https: !default_url_options[:protocol].to_s.start_with?('http:')
  )
end
```

I just shrugged the problem off as a code architecture problem and was hoping this PR here would alleviate my confusion by unifying the entire setup – but now I understood that the underlying issue was a bug elsewhere.

Tests have obviously been added in the fix, so we might skip them here. Out of curiosity and confusion, I was wondering if the test I suggested here would be green or not :) Anyway, thank you for working on this! It's greatly appreciated.

dhh commented:
I'd like to see this move forward, but I'm concerned about the backwards compatibility here. I think we should do this in two bites. First, create a new `ActionDispatch::Http::URI` class, so we don't have to munge with the existing module, which folks might have mixed in themselves in various ways. Then expose that via Rails.application.url, and derive the default_url_options for the other frameworks through that. Afterwards, once this is all good and settled, we can revisit how we might want to deprecate the existing module.

dhh commented:
Also, need to add `config.application_url` suggestions to the default env files.

aka47 commented:
Thanks, @dhh for your feedback. I added the changes you recommended, 

- I created a new ActionDispatch::Http::URI class that is used for config.application_url. 
- ActionDispatch::Http::URL is almost as it was before, some logic is in ActionDispatch::Http::URI - but the API of ActionDispatch::Http::URL is still 100% the same. In a future step, this should be refactored and some duplications removed. 
- added config.application_url to the generated environment/production|development|test.rb files

The specs are green, except for a couple of random errors in the last spec run. So I hope we are good to merge it sometime soon, finally.

zzak commented:
@aka47 Thanks for your hard work! Would you please add a changelog and squash your commits? :bow:

aka47 commented:
> @aka47 Thanks for your hard work! Would you please add a changelog and squash your commits? 🙇

True, added the changelog entries and squashed it all together. I tried to understand and fix the changelog linter, but I could not get around it. Seems like a bug.

lazaronixon commented:
Hey @aka47 on the changelog can you describe what we should do before and how it is now?

I was doing something like...

```ruby
Rails.application.configure do
  # Add top-level Rails.configuration.x.url for canonical reference
  config.x.url = URI("http://myapplication.com")

  # These options will be used as a starting point when generating URLs
  config.action_mailer.default_url_options      = { host: config.x.url.host, port: config.x.url.port }
  config.action_controller.default_url_options = { host: config.x.url.host, port: config.x.url.port }
end
```

Now, can I just do that?

```ruby
Rails.application.configure do
  # Add top-level Rails.application.url for canonical reference
  config.application_url = "http://myapplication.com"
end
```

skipkayhil commented:
> I tried to understand and fix the changelog linter, but I could not get around it. Seems like a bug.

Sorry, I'll get that fixed next week 🙇

In the meantime you can add a "." at the end of the previous entry's header and it should pass

zzak commented:
@aka47 Should be fixed now if you want to rebase :pray:

zzak commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+*   Adding `ActionDispatch::Http::URI` class.
```
```suggestion
*   Adding `ActionDispatch::Http::URI` class.

```

Should be a line break between the title and description here

zzak commented on actionpack/lib/action_dispatch.rb:
```diff
@@ -52,9 +52,10 @@ class MissingController < NameError
       autoload :Response
+      autoload :URI
     end
+    autoload :Request
```
Why was this moved?

aka47 replied:
The class is called ActionDispatch::Request and not ActionDispatch::Http::Request.
Plus, the whole http folder is a bit chaotic, this is a small first cleanup step.

zzak replied:
I got that, but I think we should not try to mix these changes as it creates unnecessary confusion. What if this introduces a regression? We have to revert this whole PR? It's better to do unrelated refactors separately IMO.

aka47 replied:
It is only the location of the file that got corrected, nothing changed that is of concern. You can still patch ActionDispatch::Reqest as you could before, not change at all.

skipkayhil replied:
> It is only the location of the file that got corrected, nothing changed that is of concern.

While the Ruby code may be equivalent, this makes git operations more difficult. As zzak mentioned, what happens if this PR needs to be reverted (or what if a change to the old file location needs to be reverted)? If bug fixes are made to the new file location, backporting them becomes more work than just cherry picking the commit. 

In general, purely cosmetic changes are rejected https://github.com/rails/rails/pull/13771#issuecomment-32746700 I'm not sure there's a strong enough reason to change this yet

aka47 replied:
I give up. You wanna argue against fixing small details, go ahead. I reverted my fix.Lets keep it a bit more chaotic as it is. And yes, git blame gets a bit more complicated, but it is just a file in the wrong place that makes it harder to understand as a developer.

aka47 replied:
sure.

zzak commented on actionpack/test/dispatch/host_authorization_test.rb:
```diff
@@ -451,7 +451,7 @@ class HostAuthorizationTest < ActionDispatch::IntegrationTest
     }
 
     assert_response :forbidden
-    assert_match "Blocked host: attacker.com#x.example.com", response.body
```
Interesting this behavior changed, I wonder if we can keep it?

aka47 replied:
Yes. I changed it back. It actually makes no difference, for some strange reasons.

zzak replied:
You mean the test passed either way? My guess is because they both will match the regex.

zzak commented on railties/CHANGELOG.md:
````diff
@@ -1,3 +1,30 @@
+
+    With this change, you only have to set it once.
+
+    ```
````
```suggestion
    ```ruby
```

zzak commented on railties/CHANGELOG.md:
````diff
@@ -1,3 +1,30 @@
+
+    ```
+    Rails.application.configure do
+        # Add top-level Rails.configuration.x.url for canonical reference
+        config.x.url = URI("http://myapplication.com")
+
+        # These options will be used as a starting point when generating URLs
+        config.action_controller.default_url_options = { host: 'myapplication.com' }
+        config.action_mailer.default_url_options     = { host: 'myapplication.com' }
````
Looks like there are 4 spaces of indentation here, it should be 2

zzak commented on railties/CHANGELOG.md:
````diff
@@ -1,3 +1,30 @@
+
+    ```
+    Rails.application.configure do
+        # Add top-level Rails.application.url for canonical reference
+        config.application_url = "http://myapplication.com"
````
Same re: indentation

zzak commented on railties/lib/rails/application/bootstrap.rb:
```diff
@@ -104,6 +105,20 @@ module Bootstrap
+          url = ActionDispatch::Http::URI.new(app.config.application_url)
+          Rails.application.url = url
+          if app.default_url_options.blank?
+            app.default_url_options = { host: url.host, protocol: url.scheme, port: url.port }
```
Since it's the same map for both conditions, how about pull it out of the branch?

zzak commented on railties/test/application/configuration_test.rb:
```diff
@@ -283,6 +283,29 @@ def change
       assert_instance_of Pathname, Rails.root
     end
 
+    test "application url is set from config file, available as Rails.application.url and sets default_url_options" do
```
Let's make this 2 separate test scenarios, long test descriptions mean having to pass all of that to `-n` when trying to run a single test

aka47 replied:
I will shorten them. To run the test separately, you should use the line number. 
```
bin/test test/application/configuration_test.rb:286
```

zzak commented on railties/test/application/configuration_test.rb:
```diff
@@ -283,6 +283,29 @@ def change
+      assert_equal({ host: "rails.test", protocol: "https", port: 443 }, Rails.application.config.action_mailer.default_url_options)
+    end
+
+    test "application url does not overwrite default_url_options nor action_mailer.default_url_options" do
```
Again, I think having separate tests for each of these is fine

zzak commented on railties/lib/rails/application/bootstrap.rb:
```diff
@@ -104,6 +105,17 @@ module Bootstrap
+          Rails.application.url = url
+          default_url_options = { host: url.host, protocol: url.scheme, port: url.port }
+          app.default_url_options = default_url_options if app.default_url_options.blank?
+          app.config.action_mailer.default_url_options = default_url_options if app.config.action_mailer.default_url_options.blank?
```
```suggestion
          app.config.action_mailer.default_url_options ||= default_url_options
```

I'm not 💯 this actually works but I wonder if it does make this line shorter?

matthewd commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,247 @@
+  mattr_accessor :tld_length, default: 1
+
+  delegate :scheme, :host, :port, :path, :query, :fragment, :to_s, to: :uri
+  delegate :scheme=, :host=, :port=, :path=, :query=, :fragment=, to: :uri
```
Absent a compelling use case, I think we should probably make this class immutable, and drop these `=` methods.

aka47 replied:
in the long term it would be nice to have a URI object for the current request, this PR is a first step, but more should follow.

matthewd replied:
Yes, I agree that's a sensible goal.. but I don't think the current request URI needs to be mutable.

aka47 replied:
Yes, we can re add it in the future if needed.

matthewd commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,247 @@
+  # number of this request is not the default HTTP \port 80 or HTTPS \port 443.
+  def port_string
+    standard_port? ? "" : ":#{port}"
+  end
```
Most of the others are more plausible, but this doesn't seem public-method-worthy to me. I'd rather not encourage callers to reconstruct a URL via string-bashing.

aka47 replied:
yes, sounds about right. Happy to change that. 
I believe there is good potential to refactor parts of the class and instance methods. And I started with it, but I reverted most of it, after feedback to do this in a 2nd step as it might be used by others and changing the API might break other people's code..

matthewd commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -1,26 +1,36 @@
-      HOST_REGEXP     = /(^[^:]+:\/\/)?(\[[^\]]+\]|[^:]+)(?::(\d+$))?/
-      PROTOCOL_REGEXP = /^([^:]+)(:)?(\/\/)?$/
+      class << self
+        def tld_length=(tld_length)
```
I think these can `delegate .., to: "ActionDispatch::Http::URI"` (unless we want to deprecate)

aka47 replied:
So, what should we do? I would deprecate those methods. Should I go ahead and deprecate those, otherwise delegate is the way to go, that makes sense..

aka47 replied:
added the delegation ..

matthewd commented on actionpack/test/dispatch/uri_test.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  test "create with ipv6" do
+    uri = ActionDispatch::Http::URI.new("http://2001:0db8:85a3:0000:0000:8a2e:0370:7334")
```
Ah, that answers my earlier question, but opens a new one: where do we expect to see such an invalid URL, and why do we care that we can parse it?

aka47 replied:
it is from the tests of the current url module

matthewd replied:
I can't find this "URI" string anywhere at the moment... got a link?

aka47 replied:
https://github.com/rails/rails/blob/main/actionpack/test/dispatch/request_test.rb#L146

matthewd replied:
Ah okay, yeah, so the input URL to produce that env should be `http://[2001:0db8:85a3:0000:0000:8a2e:0370:7334]`, the same as the `to_s` output in this assertion. It's not valid to put an IPv6 literal into a URL without the `[]` delimiters (which is why `URI.parse` isn't accepting it).

aka47 replied:
that is true, and it would be possible to remove the handling of incorrect IPv6 urls. I added them as inside the test of actionpack those incorrect urls are used and in case you want to create a URI for each request, this use case would need to be handled. For the moment, as this part got removed, we could also remove the handling of those not 100% correct urls. ( It seems they exist in the real world, or why do they got added in many of the test cases? )

matthewd commented on railties/CHANGELOG.md:
````diff
@@ -4,6 +4,32 @@
+    ```ruby
+    Rails.application.configure do
+      # Add top-level Rails.application.url for canonical reference
+      config.application_url = "http://myapplication.com"
````
Example should assume https, for most likely success if a user autopilots their copy & paste.

aka47 commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -1,26 +1,36 @@
         #    extract_domain('dev.www.example.co.uk', 2) # => "example.co.uk"
         def extract_domain(host, tld_length)
-          extract_domain_from(host, tld_length) if named_host?(host)
+          ActionDispatch::Http::URI.extract_domain(host, tld_length)
```
As URL is a module and URI a class, I believe it would be confusing and not helpful to rename URI to URL. I am happy to deprecate the old static methods for URL.

aka47 commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,247 @@
+  end
+
+  def initialize(uri_string)
+    if ipv6 = uri_string.match(IP6_REGEX)
```
no, not yet

aka47 commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -48,136 +54,25 @@ def extract_subdomain(host, tld_length)
       def initialize
         super
+        begin
+          @uri = ActionDispatch::Http::URI.new(url)
```
true, that should be done in a 2nd step.

aka47 commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -1,26 +1,36 @@
 # frozen_string_literal: true
 
 require "active_support/core_ext/module/attribute_accessors"
+require "action_dispatch/http/uri"
```
it is, without it we get an 

```
NameError: uninitialized constant ActionDispatch::Http::URI
    /workspaces/rails/actionpack/lib/action_dispatch/http/url.rb:65:
```

matthewd replied:
Is that because `bootstrap.rb` is directly requiring `url.rb` instead of requiring `action_dispatch` and then relying upon autoloads? Does the rest of the backtrace carry any clues?

Not hugely important, just moderately odd that we'd need to explicitly require something we're defining an autoload for.

matthewd commented on actionpack/lib/action_dispatch.rb:
```diff
@@ -52,8 +52,9 @@ class MissingController < NameError
       autoload :PermissionsPolicy
-      autoload :Request
       autoload :Response
+      autoload :URI
```
Oh, should this be down in `module Http` below?

matthewd commented on actionpack/lib/action_dispatch/http/uri.rb:
```diff
@@ -0,0 +1,246 @@
+require "active_support/core_ext/module/attribute_accessors"
+require "uri/generic"
+
+class ActionDispatch::Http::URI
```
I think we almost universally prefer expanded module definitions. We should probably have a cop defined for that.

aka47 committed:
use delegate and full module/class space

dhh commented:
Sorry for the slumber on this. I just got bit AGAIN from this doing a test with ActiveStorage Disk URLs generated from jobs. I know this has sat for a while, but if you're still interested in pushing it forward @aka47, I'd like to see if there's a way we could get it into Rails 8.

dhh committed:
Merge branch 'main' into add-toplevel-rails-application-url

dhh committed:
No need to change this

dhh commented on actionpack/lib/action_dispatch/routing/mapper.rb:
```diff
@@ -2033,7 +2033,7 @@ def add_route(action, controller, options, _path, to, via, formatted, anchor, op
             end
 
-            path = Mapping.normalize_path URI::RFC2396_PARSER.escape(path), formatted
+            path = Mapping.normalize_path ::URI::DEFAULT_PARSER.escape(path), formatted
```
Was this changed intended?

matthewd replied:
It needs to adopt `::URI` (because it wants to avoid our here-introduced local `URI`), but the switch to `RFC2396_PARSER` is a different recently-landed change.

```suggestion
            path = Mapping.normalize_path ::URI::RFC2396_PARSER.escape(path), formatted
```

dhh committed:
No need for this change

dhh commented:
I think we should update `ActiveStorage::Current.url` to default to `Rails.application.url` when it hasn't been set. That's another bite point. Active Storage will generate the right URL when inside the request lifecycle but not outside it (like when in a job).

dhh committed:
Update actionpack/lib/action_dispatch/routing/mapper.rb

Co-authored-by: Matthew Draper <matthewd@github.com>

dhh committed:
Explain usage, default URL to standard dev URL

dhh committed:
Use example.com as canonical URL by default in test

dhh committed:
Explain but don't set application_url

dhh commented:
Hmm, @matthewd raised a good point. We don't want URLs that before just generated paths to suddenly start generating fully formed URLs. So need to sort that as well.

jprosevear commented on railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt:
```diff
@@ -16,6 +16,10 @@ Rails.application.configure do
 
+  # Set canonical url for your application that'll be accessible via Rails.application.url.
+  # Used by Action Controller, Action Mailer, Active Storage for generating urls outside of active requests.
+  # config.application_url = "https://app.example.com"
```
Is it reasonable to have this settable with RAILS_APPLICATION_URL in the env as well?  (Maybe in a separate PR).

dhh replied:
Yeah, that's a good idea to suggest in the comment. `config.application_url = ENV["APP_URL"] || "https://app.example.com"`. And then we can have a comment in `config/deploy.yml` that corresponds to this 👍

aka47 committed:
Use RAILS_URL as Rails.application.url, and add documention about the it

Spone commented on railties/lib/rails/generators/rails/app/templates/config/deploy.yml.tt:
```diff
@@ -28,6 +28,11 @@ registry:
     - RAILS_MASTER_KEY
+
+  #   # The application url is made available as Rails.application.url and is used as the default URL for mailers and renderers.
+  #   RAILS_URL: https://app.example.com
```
To match the env var name in `production.rb.tt`

```suggestion
  #   APP_URL: https://app.example.com
```

aka47 reviewed (commented):
use env APP_URL in comment

aka47 committed:
Merge branch 'main' into add-toplevel-rails-application-url

aka47 committed:
use APP_URL

Co-authored-by: Hans Lemuet <Spone@users.noreply.github.com>

zzak commented on actionpack/lib/action_dispatch/http/url.rb:
```diff
@@ -3,179 +3,15 @@
-              yield port unless port.to_i == 80
-            end
-          end
+        delegate :tld_length, :tld_length=, :secure_protocol, :secure_protocol=,
+          :extract_domain, :extract_subdomain, :extract_subdomains, :url_for,
+          :full_url_for, :path_for, to: ActionDispatch::Http::URI
```
This is just a docs/api question, but I wonder what the purpose of having both a URL and URI class in the public domain, and if we just choose one (URL) because it's already public API than we should nodoc the URI class and keep the documentation for those public methods on the URL class (you can use `# :method: extract_domain` style rdoc notation).

aka47 replied:
As for the public API, it will be less confusing to have one URL class. The idea is to move the rdoc comments for the delegated methods to the URL class from the URI class, and then make the URI class nodoc, right?

```suggestion
# Returns the domain part of a host given the domain level.
        #   :method: extract_domain
        # 
        #    # Top-level domain example
        #    extract_domain('www.example.com', 1) # => "example.com"
        #    # Second-level domain example
        #    extract_domain('dev.www.example.co.uk', 2) # => "example.co.uk"
        delegate :tld_length, :tld_length=, :secure_protocol, :secure_protocol=,
          :extract_domain, :extract_subdomain, :extract_subdomains, :url_for,
          :full_url_for, :path_for, to: ActionDispatch::Http::URI
```

zzak commented on railties/lib/rails/generators/rails/app/templates/config/deploy.yml.tt:
```diff
@@ -33,6 +33,11 @@ registry:
+
+  #   # The application url is made available as Rails.application.url and is used as the default URL for mailers and renderers.
+  #   APP_URL: https://app.example.com
+  #   # Set number of cores available to the application on each server (default: 1).
+  #   WEB_CONCURRENCY: 2
```
```suggestion
```

zzak commented on railties/lib/rails/generators/rails/app/templates/config/environments/development.rb.tt:
```diff
@@ -15,8 +15,12 @@ Rails.application.configure do
+  # Set canonical url for your application that'll be accessible via Rails.application.url.
+  # Used by Action Controller, Action Mailer, Active Storage for generating urls outside of active requests.
+  # config.application_url = "http://localhost:3000"
+
+  # Enable/disable caching. By default caching is disabled.
+  # Run rails dev:cache to toggle caching.
```
This comment looks slightly different in the diff, maybe rebase issue?

aka47 replied:
should be fixed

zzak commented on railties/lib/rails/application/bootstrap.rb:
```diff
@@ -113,6 +114,17 @@ module Bootstrap
+          Rails.application.url = url
+          default_url_options = { host: url.host, protocol: url.scheme, port: url.port }
+          app.default_url_options = default_url_options if app.default_url_options.blank?
+          app.config.action_mailer.default_url_options = default_url_options if app.config.action_mailer.default_url_options.blank?
```
Does it matter if action_mailer is not included?

aka47 replied:
```suggestion
          if app.config.action_mailer.present? && app.config.action_mailer.default_url_options.blank?
            app.config.action_mailer.default_url_options = default_url_options
          end
```

aka47 committed:
Update development.rb.tt

fix rebase issue

aka47 committed:
Update railties/lib/rails/generators/rails/app/templates/config/deploy.yml.tt

Co-authored-by: zzak <zzakscott@gmail.com>

aka47 committed:
Merge branch 'main' into add-toplevel-rails-application-url

dhh commented:
We're going to move this to Rails 8.1. @matthewd is working on helping prepare some of the remaining work, but we're too close to the 8.0 release to get it all sorted in time.

dhh commented:
@matthewd Have you had a chance to look at this? Would love to see this land on main well in advance of 8.1 so we can really test it out.
